packages/commons/src/interop-token/interopTokenService.ts

```diff
@@ -96,11 +122,121 @@ export class InteropTokenGenerator {
       ...claims,
     };
 
-    const serializedToken = await this.createAndSignToken(
+    const serializedToken = await this.createAndSignToken({
+      header,
+      payload,
+      keyId: this.config.generatedKid,
+    });
+
+    return {
+      header,
+      payload,
+      serialized: serializedToken,
+    };
+  }
+
+  public async generateInteropApiToken({
+    sub,
+    consumerId,
+  }: {
+    sub: ClientId;
+    consumerId: TenantId;
+  }): Promise<InteropApiToken> {
+    if (
+      !this.config.generatedInteropTokenAlgorithm ||
+      !this.config.generatedInteropTokenKid ||
+      !this.config.generatedInteropTokenIssuer ||
+      !this.config.generatedInteropTokenM2MAudience ||
+      !this.config.generatedInteropTokenM2MDurationSeconds
+    ) {
+      throw Error(
+        "AuthorizationServerTokenGenerationConfig not provided or incomplete"
+      );
+    }
+
+    const currentTimestamp = Date.now();
+
+    const header: InteropJwtHeader = {
+      alg: this.config.generatedInteropTokenAlgorithm,
+      use: "sig",
+      typ: "at+jwt",
+      kid: this.config.generatedInteropTokenKid,
+    };
+
+    const payload: InteropJwtApiPayload = {
+      jti: generateId(),
+      iss: this.config.generatedInteropTokenIssuer,
+      aud: [this.config.generatedInteropTokenM2MAudience],
+      sub,
+      iat: currentTimestamp,
+      nbf: currentTimestamp,
+      exp:
+        currentTimestamp +
+        this.config.generatedInteropTokenM2MDurationSeconds * 1000,
+      [ORGANIZATION_ID_CLAIM]: consumerId,
+      [ROLE_CLAIM]: GENERATED_INTEROP_TOKEN_M2M_ROLE,
+    };
+
+    const serializedToken = await this.createAndSignToken({
+      header,
+      payload,
+      keyId: this.config.generatedInteropTokenKid,
+    });
+
+    return {
+      header,
+      payload,
+      serialized: serializedToken,
+    };
+  }
+
+  public async generateInteropConsumerToken({
+    sub,
+    audience,
+    purposeId,
+    tokenDurationInSeconds,
+  }: {
+    sub: ClientId;
+    audience: string[];
+    purposeId: PurposeId;
+    tokenDurationInSeconds: number;
+  }): Promise<InteropConsumerToken> {
+    if (
+      !this.config.generatedInteropTokenAlgorithm ||
+      !this.config.generatedInteropTokenKid ||
+      !this.config.generatedInteropTokenIssuer ||
+      !this.config.generatedInteropTokenM2MAudience
+    ) {
+      throw Error(
+        "AuthorizationServerTokenGenerationConfig not provided or incomplete"
+      );
+    }
+
+    const currentTimestamp = Date.now();
+
+    const header: InteropJwtHeader = {
+      alg: this.config.generatedInteropTokenAlgorithm,
+      use: "sig",
+      typ: "at+jwt",
+      kid: this.config.generatedInteropTokenKid,
+    };
+
+    const payload: InteropJwtConsumerPayload = {
```

Comment:
the voucher may contain also the digest

Reply:
Updated (moved the types from the `client-assertion-validation` package to models)